Calendly Block: Replace SubmitButton with new button block

aaronrobertshaw opened the pull request:
Fixes #15711, Fixes #14564 

#### Changes proposed in this Pull Request:
* Replace the SubmitButton in the Calendly block with new shared Button block.
* Fixes minor bug where Calendly spinner for inline style widget isn't hidden when there are multiple Calendly blocks on the page.
* Switch to shared Button block also fixes issue where custom text color wasn't shown in editor.

#### Prerequisites:
- Calendly account to create blocks and embed codes
- Post with Calendly blocks created prior to applying this PR's changes.
  1.   Create a post.
  2.   Add two Calendly blocks using the `link` style.
  3.   Configure one of those blocks setting its various options e.g. background color.
  4.   Add one inline Calendly block. 
  5.   Save post.
- Have the AMP plugin installed and active

#### Testing instructions:

1. On the frontend, visit post containing previous versions of Calendly block.
2. Confirm the blocks display and function as they did when you set them up. 
    _Should still be using simple `<a>` markup._
3. Edit the post.
4. Confirm the Calendly blocks still appear correctly and their settings are correct.
5. Update the link style blocks using the new settings controls provided by button block.
6. Add new Calendly block using simple URL e.g. `https://calendly.com/<username>`
7. Add new Calendly block using customized embed code for a link style widget.
8. Check that the customized attributes from the embed code have been set on the button.
9. Save the post.
10. Re-visit the post on the frontend and confirm display and function of blocks.
11. Switch back to the editor and update the embed url under the Inspector Controls > Calender Settings with a link style URL containing custom colors.
12. Confirm button colors update, save and switch to the frontend view to check there as well.
13. Click the AMP link at the top of the page and confirm links work on the AMP request view.
  ( inline style block will show as a simple link )
14. Run tests in `extensions/blocks/calendly/test/utils.js`

<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->
#### Before:
<img width="1084" alt="Calendly-Before-SubmitButtonReplacement" src="https://user-images.githubusercontent.com/60436221/82451079-dbe0e900-9af0-11ea-90c7-fe15ff918514.png">

#### After:
<img width="1066" alt="Calendly-After-SubmitButtonReplacement" src="https://user-images.githubusercontent.com/60436221/82451104-e26f6080-9af0-11ea-9ee5-5ba85525ca92.png">


#### Proposed changelog entry for your changes:
* Update Calendly block to use new shared Button block.

aaronrobertshaw committed:
Replace Calendly SubmitButton with shared button

The new shared button offers a few extra features and an oportunity to make
multiple components more consistent. As such, the SubmitButton in the Calendly
block is being replaced.

aaronrobertshaw committed:
Clean up call to migrateAttributes

Copons commented on extensions/blocks/calendly/deprecated/v1/index.js:
```diff
@@ -0,0 +1,107 @@
+		return [ newAttributes, newInnerBlocks ];
+	},
+	isEligible: ( attributes, innerBlocks ) =>
+		isEmpty( innerBlocks ) || some( pick( attributes, deprecatedAttributes ), Boolean ),
```
In https://github.com/Automattic/jetpack/pull/15918 I've also made sure to only migrate when the block is set to display the button, which in this case means `style === 'link'`, while leaving the embed style alone.

```js
	isEligible: ( attributes, innerBlocks ) =>
		'link' === attributes.style &&
		( isEmpty( innerBlocks ) || some( pick( attributes, deprecatedAttributes ), Boolean ) ),
```

Not a big deal, but I think we'd have a tiny performance improvement.

aaronrobertshaw replied:
Still an easy improvement to make, so I've tightened that up thanks.

Copons commented on extensions/blocks/calendly/index.js:
```diff
@@ -40,15 +50,15 @@ export const settings = {
 			style: 'inline',
 			url: CALENDLY_EXAMPLE_URL,
 		},
+		innerBlocks: [ innerButtonBlock ],
```
The example is `style: 'inline'`, which is the actual widget embed, where the Button is not displayed.
So, defining the `innerBlocks` here is unnecessary.

This would also slightly simplify the how we pass around data.
We can avoid definining and exporting `innerButtonBlock` here in the index.js, and instead use it straight away in edit.js.

aaronrobertshaw replied:
Good catch. I've cleaned that up and moved it to edit.js.

Copons commented on extensions/blocks/calendly/calendly.php:
```diff
@@ -86,35 +80,27 @@ function load_assets( $attr, $content ) {
+			$content = preg_replace(
+				'/data-id-attr="placeholder"/',
+				'id="' . esc_attr( $block_id ) . '"',
+				$content
```
I'd recommend to do a simpler thing here (again, inspired by what I did in #15918).

- Give a `uniqueId` attribute to the Button inner block, e.g. `uniqueId: 'calendly-widget-id'`.
- Do a simple string replace here: `$content = str_replace( 'calendly-widget-id', esc_attr( $block_id ), $content );`

(good call about the `esc_attr`! will need to update the Eventbrite PR! 🙂 )

aaronrobertshaw replied:
Thanks, I've simplified that as per your suggestion 🙂

Copons commented:
This works well! @aaronrobertshaw 
I've left a few minor comments, and then I've got a major one below. 😅 

Just one thing:

> Fixes minor bug where Calendly spinner for inline style widget isn't hidden when there are multiple Calendly blocks on the page.

This is happening to me regardless. 🤔 
Tested with two Calendly blocks: a widget and a button.

---

I've took the liberty of assigning you to https://github.com/Automattic/jetpack/issues/15498, which is directly related to this PR.

As it is right now, the button doesn't work in AMP.
(You can try it yourself with the [AMP plugin](https://en-gb.wordpress.org/plugins/amp/), previewing the post with the AMP button).

The idea proposed in #15498 was to replace the button with a link when in AMP.
But, looking at the Calendly block, I don't see any good reasons to render the Button as a `button`.
Instead, we could set it as an `element: 'a'` and pass the Calendly link to the `url` attribute.
In AMP it should just work as-is, while we could simply add a `event.preventDefault()` when we attach the click handler.
Having the Button as a link, it wouldn't be stripped away by the WPCOM KSES, and so we could even go as far as to save in the post content (handy Button `saveInPostContent` attribute 😄).

There is a gotcha though: to pass the Calendly link from the parent block to the child Button, and keep the value in sync, we need a small update to the Button block, based upon an [old discussion](https://github.com/Automattic/jetpack/pull/15370#issuecomment-614071023).
I'm opening the PR as we speak!

aaronrobertshaw committed:
Improve when to migrate deprecated Calendly block

aaronrobertshaw committed:
Remove unnecessary deffinition of innerButtonBlock

The example is style: 'inline', which is the actual widget embed, where the Button is not displayed. So, defining the innerBlocks here is unnecessary. We can just define it within the edit.js where it is used instead.

aaronrobertshaw committed:
Simplify setting unique id for block markup

aaronrobertshaw committed:
Switch inner button block to render link

aaronrobertshaw committed:
Allow new embed URLs from sidebar to update button attrs

aaronrobertshaw committed:
Fall back to link for inline style for AMP requests

aaronrobertshaw commented:
Thanks @Copons, appreciate all the suggestions and direction on this one. Especially the AMP side of things, I was a little at a loss on how to test that.

> But, looking at the Calendly block, I don't see any good reasons to render the Button as a button. Instead, we could set it as an element: 'a' and pass the Calendly link to the url attribute.
In AMP it should just work as-is, while we could simply add a event.preventDefault() when we attach the click handler.

I agree regarding this block only needing to render as a link. I've updated it accordingly and also tweaked it to pass those attributes through to the button block. Happy to come back, rebase this and update to use the `passthroughAttributes` from #15931 once it's merged or as a separate PR.

While making these updates I found I'd missed updating the button block's attributes when the embed URL from the sidebar calendar settings section was changed and contained embedded styles. These sound like another candidate to use the `passthroughAttributes` with as well.

The buttons work ok for me now via AMP although if there are no custom colors used to provide inline styles, their appearance could be sketchy. The block's inline style, that embeds an iframe, obviously wasn't working either, so I've made that just fallback to a simple link. Do you think it's worth just making the block render a simple link for all AMP requests to avoid styling inconsistencies?

>>Fixes minor bug where Calendly spinner for inline style widget isn't hidden when there are multiple Calendly blocks on the page.

>This is happening to me regardless. 🤔
Tested with two Calendly blocks: a widget and a button.

I might not have been clear on that one, sorry. The spinner was just appearing outside the bounds of the block's container for me. The fix was just adding a simple `position: relative` on that inline style container so its positioning appeared ok until it was eventually covered by the iframed widget.

Before: 
![calendly-spinner](https://user-images.githubusercontent.com/60436221/82996403-4097c880-a048-11ea-9b54-2b0d55042b99.gif)

After:
<img width="623" alt="Screen Shot 2020-05-27 at 6 29 18 pm" src="https://user-images.githubusercontent.com/60436221/82996419-45f51300-a048-11ea-8626-93984207a30d.png">

Can you confirm it's still an issue?

Copons commented:
As mentioned internally, I've noticed that for some reasons using the Button with `InnerBlocks` is preventing the Calendly block from being selected.
I'm not sure why this happens, but I've noticed that it's a regression from one of last night / yesterday commits, since in 265ec83 blocks are selectable.

---

> The spinner was just appearing outside the bounds of the block's container for me. The fix was just adding a simple position: relative on that inline style container so its positioning appeared ok until it was eventually covered by the iframed widget.

I see, and yes, the spinner is not in that awkward position anymore.
Though, now it's below the initial state of the widget:
<img width="679" alt="Screenshot 2020-05-27 at 15 01 09" src="https://user-images.githubusercontent.com/2070010/83037470-b8a7c380-a033-11ea-9127-7b8c1976b0ff.png">
It never disappears, so it's not clear what it does.
Though, when we click on an event, the spinner gets covered.

I guess we could reposition it where it would _always_ be covered by the widget.
E.g. we could override the `.calendly-spinner{ top: 50% }` (set by the widget's own CSS) to something like `top: 50px`:

<img width="613" alt="Screenshot 2020-05-27 at 16 09 38" src="https://user-images.githubusercontent.com/2070010/83038274-ad08cc80-a034-11ea-9496-815ae354d755.png">

(In the screenshot I've also changed the z-index to make it appear above the widget)

aaronrobertshaw committed:
Relocate calendly spinner

Given the varying heights of the inline styled calendly blocks, its possible that the spinner is still visible after the widget loads. This will reposition it so it gets hidden.

aaronrobertshaw committed:
Prevent full-width block triggering button popup

Previously, on the frontend, clicking in the empty space beside the button would attempt to trigger the button's click.

aaronrobertshaw committed:
Fix inconsistent block id placement and event listener

Previous fix to event listener solved deperecated markup clicks but created issue on migrated and new blocks.

aaronrobertshaw commented:
Thanks for highlighting the issue with the spinner, I hadn't tried the inline widget with fewer than 2 meeting durations. I've tidied that up.

> As mentioned internally, I've noticed that for some reasons using the Button with InnerBlocks is preventing the Calendly block from being selected.
I'm not sure why this happens, but I've noticed that it's a regression from one of last night / yesterday commits, since in 265ec83 blocks are selectable.

I'm not sure I follow regarding the Calendly blocks not being selectable. In the editor, for me, clicking the blank space beside the button block selects the Calendly block itself. Alternatively, moving between the blocks with the arrow keys still lets me select the Calendly block.

On the frontend, I found an inconsistency with which element the block ID was added to, between the deprecated and current versions of the link style blocks. I've fixed that up.

Hopefully, that sorts out the issue.

Copons commented:
@aaronrobertshaw Apparently the broken selection only happens with the Guten plugin activated. Tested with 8.1 and 8.2 (freshly checked out this morning).

![May-28-2020 18-40-32](https://user-images.githubusercontent.com/2070010/83174911-ff201f80-a112-11ea-9830-d9e7c58cbdae.gif)

Copons committed:
Restore inner block in example to prevent the block being unselectable

Copons commented:
@aaronrobertshaw I've finally run a `git bisect` and found what caused the selection issue (but not _why_).
The bad commit was bd30818 (Remove unnecessary definition of innerButtonBlock), so my best guess is that on recent Gutens, if you don't add the inner blocks to the example, the example will break (silently!), and the sidebar can't open, messing up the selection.
I might be wrong, but I've pushed fdedc24 which fixes the issue for me.

aaronrobertshaw commented:
@Copons Thanks for taking the time to look into that one. I understand the issue you were highlighting now, the new commit also has it working properly for me too.

Copons commented:
Just a heads up that I've merged #15931 (the `passthroughAttributes` one)!

aaronrobertshaw committed:
Use new passthroughAttributes to update button URL

Only the URL has been added to passthrough attributes as other properties that may set via embed code can be updated via the button's block settings.

aaronrobertshaw committed:
Add CSS style to fix center alignment for button

aaronrobertshaw commented:
I've rebased this PR as well as switched it to use the new `passthroughAttributes` for the inner button block's URL. 

There is still the ability for users to update the embed URL from the Calendly block's "Calendar Settings" in the sidebar. These embed URLs can contain custom styles that should be applied to the button. Given this isn't only a one way syncing of the attribute, like the URL, I've left the code that will update the button attributes after parsing a new embed URL.

Also, added style to fix center alignment of inner button block on frontend.

Copons commented on extensions/blocks/calendly/view.scss:
```diff
@@ -10,3 +10,7 @@
 }
+
+.wp-block-jetpack-calendly.aligncenter .wp-block-jetpack-button {
+	text-align: center;
```
I think we should also center it in the editor.

aaronrobertshaw replied:
@copons Blocks with the button center-aligned are showing fine for me in the editor due to to the following CSS:

```
.block-editor-block-list__block[data-align=center] {
     text-align: center;
}
```

Are you aware of a case where this is insufficient?

Copons replied:
Apologies, I should have added more context.
I've tested (and reproduced again today) with the Gutenberg plugin active (v8.3.0).

I can also confirm that the button centers correctly when the plugin is deactivated.

Copons replied:
Apparently the block alignment behaviour changed slightly.
Instead of applying the `data-align` to the block div, it now creates a wrapper `.wp-block` div with the `data-align`.

aaronrobertshaw replied:
Thanks for pointing that one out. I've tweaked the styles and its working for me with the Gutenberg plugin (8.3.0) installed and active as well as without it.

Copons reviewed (dismissed):
This works well for me!

I've left a comment for a minor fix, but otherwise this LGTM! ✨

aaronrobertshaw committed:
Fix editor center alignment for button

With the Gutenberg plugin v8.3.0 installed and active it creates slightly different markup so the preexisting styles no longer center align the button. This fixes that and improves frontend style.

Copons reviewed (dismissed):
Nice, thanks @aaronrobertshaw!

jeherve commented on extensions/blocks/calendly/calendly.php:
```diff
@@ -160,3 +152,85 @@ function get_attribute( $attributes, $attribute_name ) {
+		'jetpack-calendly-external-js',
+		'https://assets.calendly.com/assets/external/widget.js',
+		null,
+		JETPACK_VERSION,
```
Quick change, it should be `JETPACK__VERSION` here.

```suggestion
		JETPACK__VERSION,
```

jeherve reviewed (changes requested):
In my tests, I seem to be losing the color of the existing buttons when I update to this branch.

### Steps to reproduce

1. On `master`, create a new Calendly block with a button style
2 Give a custom background color to your button, something that's not among the predefined options.
3. Publish your post.
4. Switch to this branch.
5. On the frontend of your site, the button color is back to the default.

```
<div class="wp-block-jetpack-calendly aligncenter calendly-style-link"><a id="calendly-block-2" class="wp-block-button__link has-background" href="https://href.li/?https://calendly.com/jeherve?hide_event_type_details=0&amp;background_color=ffffff&amp;text_color=4D5055&amp;primary_color=00A2FF" role="button" rel="noreferrer">Click to book a time slot</a></div>
```

6. In the editor, the color is visible.

```
<div aria-label="Block: Button" role="group" id="block-77202acd-6fb7-4965-b6be-04819eb3a9b8" class="block-editor-block-list__block wp-block" data-block="77202acd-6fb7-4965-b6be-04819eb3a9b8" data-type="jetpack/button" data-title="Button" tabindex="0" style="transform-origin: center center;"><div> <div class="wp-block-button wp-block-jetpack-button"><div role="textbox" aria-multiline="true" aria-label="Add text…" class="rich-text block-editor-rich-text__editable wp-block-button__link has-background" contenteditable="true" style="background-color: rgb(6, 172, 9); white-space: pre-wrap;">Click to book a time slot</div></div> </div></div>
```

7. Opening that post in the editor triggers the migration and we get our color back:
```
<a class="wp-block-button__link has-background" style="background-color: #06ac09;" data-id-attr="calendly-block-2" id="calendly-block-2" href="https://href.li/?https://calendly.com/jeherve" target="_blank" role="button" rel="noopener noreferrer noreferrer">Click to book a time slot</a>
```

Is there any way to avoid that?


******

Other than that, everything seems to work well in my tests.

fgiannar commented:
Please note the following as well:

I tried to add a new Calendly block using customized embed code with custom page settings, like:
https://d.pr/i/3ls3Vt

The custom page settings were lost. 
Seems that the button `href` attribute was missing the query string : `?background_color=212020&text_color=801249&primary_color=19ff00`

aaronrobertshaw committed:
Fix JETPACK__VERSION constant use

Co-authored-by: Jeremy Herve <jeremy@jeremy.hu>

aaronrobertshaw committed:
Fix styles for deprecated button styles

aaronrobertshaw committed:
Update block button url with embed params

After switch to using the block button and passing through the url attribute, the embed params for background, text and primary color were lost. The original approach to parsing embed code strips the url back to basics and only adds the query string params when generating markup.

This change simply replaces the base url with the one containing query string params as this had the least impact on existing approach.

aaronrobertshaw committed:
Set button blocks default font color

After the switch to the shared button block, the default color provided by the .wp-block wrapper element was lost this sets the same color.

aaronrobertshaw commented:
The custom styles are now correctly applied to the deprecated button element and custom page settings for embed codes have been fixed.

I also added a style to provide the default font color, that a wrapping `.wp-block-button` did before the switch to the shared button component. This fixes up an issue where buttons with visited URLs inherited a black font color.

fgiannar commented:
Thanks for the fixes @aaronrobertshaw 
Tested both for @jeherve and my comments as well and looks fine!

jeherve commented:
r210102-wpcom
